sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala

```diff
@@ -0,0 +1,112 @@
+import org.apache.spark.sql.catalyst.catalog.{CatalogStatistics, HiveTableRelation}
+import org.apache.spark.sql.catalyst.expressions.{And, AttributeReference, AttributeSet, Expression, PredicateHelper, SubqueryExpression}
+import org.apache.spark.sql.catalyst.planning.PhysicalOperation
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.command.CommandUtils
+
+/**
+ * TODO: merge this with PruneFileSourcePartitions after we completely make hive as a data source.
+ */
+private[sql] class PruneHiveTablePartitions(session: SparkSession)
+  extends Rule[LogicalPlan] with PredicateHelper {
+  /**
+   * Extract the partition filters from the filters on the table.
+   */
+  private def extractPartitionPruningFilters(
+              filters: Seq[Expression],
+              relation: HiveTableRelation): Seq[Expression] = {
+    val normalizedFilters = filters.map { e =>
+      e transform {
+        case a: AttributeReference =>
+          a.withName(relation.output.find(_.semanticEquals(a)).get.name)
+      }
+    }
+    val partitionSet = AttributeSet(relation.partitionCols)
+    normalizedFilters.filter { predicate =>
+      !predicate.references.isEmpty && predicate.references.subsetOf(partitionSet)
+    }
+  }
+
+  /**
+   * Prune the hive table using filters on the partitions of the table,
+   * and also update the statistics of the table.
+   */
+  private def prunedHiveTableWithStats(
+              relation: HiveTableRelation,
+              partitionFilters: Seq[Expression]): HiveTableRelation = {
+    val conf = session.sessionState.conf
+    val prunedPartitions = session.sessionState.catalog.listPartitionsByFilter(
+      relation.tableMeta.identifier,
+      partitionFilters)
+    val sizeInBytes = try {
+      val sizeOfPartitions = prunedPartitions.map { partition =>
+        val rawDataSize = partition.parameters.get(StatsSetupConst.RAW_DATA_SIZE).map(_.toLong)
+        val totalSize = partition.parameters.get(StatsSetupConst.TOTAL_SIZE).map(_.toLong)
+        if (rawDataSize.isDefined && rawDataSize.get > 0) {
+          rawDataSize.get
+        } else if (totalSize.isDefined && totalSize.get > 0L) {
+          totalSize.get
+        } else {
+          CommandUtils.calculateLocationSize(
+            session.sessionState, relation.tableMeta.identifier, partition.storage.locationUri)
+        }
+      }.sum
+      // If size of partitions is zero fall back to the default size.
+      if (sizeOfPartitions == 0L) conf.defaultSizeInBytes else sizeOfPartitions
+    } catch {
+      case e: IOException =>
+        logWarning("Failed to get table size from HDFS.", e)
+        conf.defaultSizeInBytes
+    }
+    val newTableMeta =
+      if (relation.tableMeta.stats.isDefined) {
+        relation.tableMeta.copy(
+          stats = Some(relation.tableMeta.stats.get.copy(sizeInBytes = BigInt(sizeInBytes))))
```

Comment:
is it safe to keep other stats after we prune partitions?

Reply:
could be inconsistent, eg. rowCount and sizeInBytes is inconsistent.
Restored to new CatalogStatistics instance. But by doing so, some statistics may be lost which should not impact accuracy.